_docs/4-kubernetes/3-bookinfo.md

```diff
@@ -0,0 +1,200 @@
+
+   ```bash
+   $ kubectl describe pod gateway-j6vr6 | grep Node
+   Node:		node4.k8s.example.com/9.147.221.77
+   $ export GATEWAY_URL=9.147.221.77:32000
+   ```
+
+### Content Based Routing
+
+1. Set the default routes to v1 of each service
+
+   ```bash
+   $ kubectl create -f examples/k8s-bookinfo-default-routes.yaml
+   routingrule "set-productpage-v1" created
+   routingrule "set-details-v1" created
+   routingrule "set-ratings-v1" created
+   routingrule "set-reviews-v1" created
+   ```
+
+   You can confirm the routes are defined and valid by inspecting the rules' `Status` field returned in:
+
+   ```bash
+   $ kubectl get routingrules -o json
+   ```
+
+   If you open the Bookinfo URL (`http://$GATEWAY_URL/productpage/productpage`) in your browser,
+   you should see the bookinfo application `productpage` displayed. Notice that the `productpage`
+   is displayed, with no rating stars since `reviews:v1` does not access the ratings service.
+
+1. Route a specific user to `reviews:v2`
+
+   Lets enable the ratings service for test user "jason" by routing productpage traffic to
+   `reviews:v2` instances.
+
+   ```bash
+   $ kubectl create -f examples/k8s-bookinfo-jason-reviews-v2-route-rules.yaml
+   routingrule "set-reviews-v2-user-jason" created
+   ```
+
+   Confirm the rule is applied and valid:
+
+   ```bash
+   $ kubectl get routingrule set-reviews-v2-user-jason -o json
+   {
+      "apiVersion": "amalgam8.io/v1",
+      "kind": "RoutingRule",
+      ...
+      "status": {
+         "state": "valid"
+      }
+   }
+   ```
+
+   Log in as user "jason" at the `productpage` web page. You should now see ratings (1-5 stars) next
+   to each review.
+
+### Resiliency Testing
+
+   To test our bookinfo application microservices for resiliency, we will _inject a 7s delay_
+   between the reviews:v2 and ratings microservices. Since the _reviews:v2_ service has a
+   10s timeout for its calls to the ratings service, we expect the end-to-end flow to
+   continue without any errors.
+
+1. Inject the delay
+
+   Create a fault injection, to delay traffic coming from user "jason" (our test user).
+
+   ```bash
+   $ kubectl create -f examples/k8s-bookinfo-jason-7s-delay.yaml
+   routingrule "set-delay-user-jason" created
+   ```
+
+   Verify that the rule has been set:
+
+   ```bash
+   $ kubectl get routingrule set-delay-user-jason -o json | sed -n '/status/,/}/p'
+      "status": {
+         "state": "valid"
+      }
+   ```
+
+1. Confirm application behavior
+
+   If the application's front page was set to correctly handle delays, we expect it
+   to load within approximately 7 seconds. To see the web page response times, open the
+   *Developer Tools* menu in IE, Chrome or Firefox (typically, key combination _Ctrl+Shift+I_
+   or _Alt+Cmd+I_) and reload the `productpage` web page.
+
+   You will see that the webpage loads in about 6 seconds. The reviews section will show
+   *Sorry, product reviews are currently unavailable for this book*.
+
```

Comment:
The reason that the entire reviews service has failed is because our bookinfo application has a bug. The timeout between the productpage and reviews service is less (3s) than the timeout between the reviews and ratings service (10s). These kinds of bugs can occur in typical enterprise applications where different teams develop different microservices independently.